zipkin/src/main/java/zipkin/internal/Node.java

```diff
@@ -132,6 +132,8 @@ public void remove() {
     Map<Long, Long> idToParent = new LinkedHashMap<>(idToNode.size());
 
     public void addNode(@Nullable Long parentId, long id, V value) {
+      // drop circular deps vs NPE later
+      if (parentId != null && parentId.equals(id)) return;
```

Comment:
I'll change this code to throw, and the code that calls it to skip and log (since that code has the logger people turn on)